File: 백준/Gold/3020. 개똥벌레/개똥벌레.java
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.StreamTokenizer;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

class Main {
    static StreamTokenizer sttk = new StreamTokenizer(new InputStreamReader(System.in));
    static List<Integer> downToUp;
    static List<Integer> upToDown;

    static int readInt() throws IOException {
        sttk.nextToken();
        return (int) sttk.nval;
    }

    public static void main(String[] args) throws IOException {
        int n = readInt();
        int h = readInt();

        downToUp = new ArrayList<>(n >>> 1);
        upToDown = new ArrayList<>(n >>> 1);

        for (int i = 0; i < (n >>> 1); i++) {
            downToUp.add(readInt());
            upToDown.add(readInt());
        }

        downToUp.sort(Comparator.naturalOrder());
        upToDown.sort(Comparator.naturalOrder());

        int min = n;
        int count = 0;

        for (int i = 1; i <= h; i++) {
            int sum = bisect(upToDown, h - i + 1) + bisect(downToUp, i);

            if (sum < min) {
                min = sum;
                count = 1;
            } else if (sum == min) {
                count++;
            }
        }

        System.out.printf("%d %d", min, count);
    }

    static int bisect(List<Integer> list, int v) {
        int ret = list.size();
        int s = 0;
        int e = list.size() - 1;

        while (s <= e) {
            int mid = (s + e) >>> 1;

            if (list.get(mid) < v) {
                s = mid + 1;
            } else {
                e = mid - 1;
                ret = mid;
            }
        }

        return list.size() - ret;
    }
}

File: 백준/Gold/3020. 개똥벌레/개똥벌레.kt
import java.io.StreamTokenizer

const val MAX_VALUE = 500_000
val sttk = StreamTokenizer(System.`in`.bufferedReader())

fun readInt() = with(sttk) {
    nextToken()
    nval.toInt()
}

fun main() {
    val n = readInt()
    val h = readInt()

    val DU = IntArray(MAX_VALUE + 1)
    val UD = IntArray(MAX_VALUE + 1)

    repeat(n ushr 1) {
        DU[readInt()]++
        UD[h - readInt() + 1]++
    }

    for (i in 1..h) {
        UD[i] += UD[i - 1]
        DU[h - i] += DU[h - i + 1]
    }

    var min = n
    var count = 0

    for (i in 1..h) {
        val sum = DU[i] + UD[i]

        if (sum < min) {
            min = sum
            count = 1
        } else if (sum == min) {
            count++
        }
    }

    print("$min $count")
}